File: 2019-11-11/Nicitis/2번문제.cpp
#include <iostream>
using namespace std;

int solution(int n)
{
	/* 풀이법 1. 재귀함수
	n칸 이동하는데 소모되는 최소 배터리 소모량 f(n)이라 할 때,
	n이 홀수일 때, f(n) = f(n - 1) + 1
	n이 짝수일 때, f(n) = f(n / 2)
	*/

	if (n == 1)
		return 1;
	else if (n % 2 == 1)
		return solution(n - 1) + 1;
	else
		return solution(n / 2);

	/* 풀이법 2. 나머지
	f(n)은 n이 0이 될 때까지 2로 나눌 때, 나머지의 합과 같다.
	for문으로 풀 수 있다. 위와 같은 원리.
	*/
}

File: 2019-11-18/heeiui/2번문제.java
import java.util.StringTokenizer;
import java.util.HashMap;

class Solution {
	public String[] solution(String[] record) {
		String[] answer = {};
		int ansCnt = 0; 
		int ansIdx = 0;
		int cmdCnt = record.length;
		StringTokenizer st;
		HashMap<String, String> ID = new HashMap<String, String>();
		
		for(int i = 0; i < cmdCnt; i++)
        {
			st = new StringTokenizer(record[i], " ");
			String cmd = st.nextToken();
			String userID = st.nextToken();
			String NickName;
			
			if(cmd.equals("Enter") || cmd.equals("Change")) 
            { //Enter 또는 Change일 경우 ID에 userID와 NickName을 맵핑
				NickName = st.nextToken();
				ID.put(userID, NickName);
			}
			
			if(cmd.equals("Enter") || cmd.equals("Leave"))
            { 
				ansCnt++;
            }
		}
			answer = new String[ansCnt];
    
		for(int i = 0; i < cmdCnt; i++) 
        {
			st = new StringTokenizer(record[i], " ");
			String cmd = st.nextToken();
			String userID = st.nextToken();
			String NickName;
			
			if(cmd.equals("Enter")) 
            {
				NickName = ID.get(userID);
				answer[ansIdx++] = NickName + "님이 들어왔습니다.";
			}
			else if(cmd.equals("Leave")) 
            {
				NickName = ID.get(userID);
				answer[ansIdx++] = NickName + "님이 나갔습니다.";
			}
		}
		
		return answer;
    }
}
